presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1116,34 +1116,76 @@ protected Scope visitJoin(Join node, Optional<Scope> scope)
             Scope left = process(node.getLeft(), scope);
             Scope right = process(node.getRight(), isLateralRelation(node.getRight()) ? Optional.of(left) : scope);
 
-            Scope output = createAndAssignScope(node, scope, left.getRelationType().joinWith(right.getRelationType()));
-
-            if (node.getType() == Join.Type.CROSS || node.getType() == Join.Type.IMPLICIT) {
-                return output;
-            }
-
             if (criteria instanceof JoinUsing) {
-                // TODO: implement proper "using" semantics with respect to output columns
                 List<Identifier> columns = ((JoinUsing) criteria).getColumns();
 
-                List<Expression> expressions = new ArrayList<>();
+                List<Field> joinFields = new ArrayList<>();
+
+                List<Integer> leftJoinFields = new ArrayList<>();
+                List<Integer> rightJoinFields = new ArrayList<>();
+
+                Set<Identifier> seen = new HashSet<>();
                 for (Identifier column : columns) {
-                    Expression leftExpression = new Identifier(column.getValue());
-                    Expression rightExpression = new Identifier(column.getValue());
+                    if (seen.contains(column)) {
+                        throw new SemanticException(DUPLICATE_COLUMN_NAME, column, "Column '%s' appears multiple times in USING clause", column.getValue());
+                    }
+                    seen.add(column);
+
+                    QualifiedName columnName = QualifiedName.of(column.getValue());
+                    Optional<ResolvedField> leftField = left.tryResolveField(column, columnName);
+                    Optional<ResolvedField> rightField = right.tryResolveField(column, columnName);
+
+                    if (!leftField.isPresent() || !rightField.isPresent()) {
+                        throw missingAttributeException(column, columnName);
```

Comment:
maybe we could say which relation is missing this column. What do you think?

Reply:
Yes, good idea